compiler/rustc_borrowck/src/diagnostics/mod.rs

```diff
@@ -1133,8 +1132,45 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                                 place_name, partially_str, loop_message
                             ),
                         );
+                        let ty = moved_place.ty(self.body, self.infcx.tcx).ty;
+                        if let ty::Adt(def, ..) = ty.kind()
+                            && Some(def.did()) == self.infcx.tcx.lang_items().pin_type()
+                        {
+                            err.span_suggestion_verbose(
+                                fn_call_span.shrink_to_lo(),
+                                "consider reborrowing the `Pin` instead of moving it",
+                                "as_mut().".to_string(),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                        if let Some(clone_trait) = tcx.lang_items().clone_trait() {
+                            // We can't use `predicate_may_hold` or `can_eq` without ICEs in
+                            // borrowck because of the inference context, so we do a poor-man's
+                            // version here.
+                            for impl_def_id in tcx.all_impls(clone_trait) {
+                                if let Some(def_id) = impl_def_id.as_local()
+                                    && let hir_id = tcx.hir().local_def_id_to_hir_id(def_id)
+                                    && let hir::Node::Item(hir::Item {
+                                        kind: hir::ItemKind::Impl(hir::Impl {
+                                            self_ty,
+                                            ..
+                                        }),
+                                        ..
+                                    }) = tcx.hir().get(hir_id)
+                                {
+                                    if ty == hir_ty_to_ty(tcx, self_ty) {
```

Comment:
I think you can use `tcx.type_of(impl_def_id)` instead of invoking astconv itself.

Reply:
`hir_ty_to_ty` is ["quasi-deprecated"](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir_analysis/lib.rs.html#539), so it's preferential not to use it if possible.

Reply:
Wouldn't `tcx.type_of(impl_def_id)` be the type of the `impl` itself? Otherwise I can peel refs and compare the `DefId` of the types and accomplish the same, I think.

Reply:
That query returns It's the `Self` type of the impl, which is what you want, right?